File: sites/exp.arolariu.ro/telemetry/health_policy.py
"""Health and connectivity telemetry suppression policy for the exp service.

Sole owner of the suppressed path list. Consumed by the request middleware in ``main``
to keep probe traffic out of exported logs, and by ``TelemetrySettings`` to build the
trace/HTTP-metric exclusion patterns handed to ``FastAPIInstrumentor``.
"""

import os
import re

SUPPRESSED_HEALTH_PATHS: tuple[str, ...] = ("/health", "/api/health", "/api/ready")

SUPPRESSION_ENV_VAR: str = "OTEL_SUPPRESS_HEALTH_TELEMETRY"


def _normalize(path: str) -> str:
    """Strip the query string and any trailing slash from a request path."""

    without_query = path.split("?", 1)[0]
    return without_query.rstrip("/") if len(without_query) > 1 else without_query


def parse_suppression_flag(raw_value: str | None) -> bool:
    """Return whether suppression is enabled for the supplied raw environment value.

    Unset and unparseable values resolve to the safe default of enabled suppression;
    only the literal string ``false`` disables it.
    """

    if raw_value is None or not raw_value.strip():
        return True

    return raw_value.strip().lower() != "false"


def is_suppressed_path(path: str | None) -> bool:
    """Return whether ``path`` is one of the suppressed health endpoints.

    Matching is exact on the normalized path and case-insensitive. Prefix matching is
    deliberately avoided so that paths such as ``/healthcheck-admin`` are not swallowed.
    """

    if not path:
        return False

    return _normalize(path).lower() in SUPPRESSED_HEALTH_PATHS


def should_suppress_telemetry(path: str | None) -> bool:
    """Return whether telemetry must be suppressed for ``path``.

    Fails safe toward suppression: an unset or malformed override leaves suppression
    enabled, so cost control holds even when the environment is misconfigured. Only the
    literal string ``false`` re-enables health telemetry.
    """

    return parse_suppression_flag(os.environ.get(SUPPRESSION_ENV_VAR)) and is_suppressed_path(path)


def build_excluded_urls() -> str:
    """Return the comma-separated exclusion patterns for ``FastAPIInstrumentor``.

    OpenTelemetry's ``ExcludeList`` joins the supplied entries with ``|`` and applies
    ``re.search``, so a bare path such as ``/api/health`` is an **unanchored** regex that
    would also swallow ``/api/healthy`` and ``/api/health/extra``. That contradicts the
    exact-match policy implemented by :func:`is_suppressed_path`.

    Each path is therefore emitted as an anchored, case-insensitive expression tolerating an
    optional trailing slash and an optional query string, so trace and HTTP-metric exclusion
    matches :func:`is_suppressed_path` exactly.
    """

    return ",".join(f"(?i:^{re.escape(path)}/?(\\?.*)?$)" for path in SUPPRESSED_HEALTH_PATHS)

File: sites/arolariu.ro/src/lib/telemetry/healthPolicy.ts
/**
 * @fileoverview Health and connectivity telemetry suppression policy.
 * @module sites/arolariu.ro/src/lib/telemetry/healthPolicy
 *
 * @remarks
 * Sole owner of the suppressed path list for the website. Consumed by the OpenTelemetry
 * ignore hooks in `instrumentation.server.ts`. Deliberately free of `server-only` so it
 * remains directly unit-testable.
 */

/** Paths whose telemetry is suppressed. Matched exactly, never by prefix. */
export const SUPPRESSED_HEALTH_PATHS: readonly string[] = ["/health", "/api/health", "/api/ready"];

/** Environment variable controlling suppression. Only the literal `false` disables it. */
export const SUPPRESSION_ENV_VAR = "OTEL_SUPPRESS_HEALTH_TELEMETRY";

/**
 * Parses the raw suppression environment variable value.
 * @param rawValue The raw environment variable value, possibly undefined or malformed.
 * @returns `false` only when the value is exactly `false`, case-insensitively.
 */
export function parseSuppressionFlag(rawValue: string | undefined): boolean {
  if (rawValue === undefined || rawValue.trim() === "") return true;
  return rawValue.trim().toLowerCase() !== "false";
}

function normalize(path: string): string {
  const queryIndex = path.indexOf("?");
  const withoutQuery = queryIndex === -1 ? path : path.slice(0, queryIndex);
  if (withoutQuery.length <= 1) return withoutQuery;

  // Trailing slashes are trimmed by scanning rather than with /\/+$/u: the quantifier in
  // that pattern backtracks super-linearly on adversarial input (sonarjs/super-linear-regex).
  let end = withoutQuery.length;
  while (end > 0 && withoutQuery[end - 1] === "/") end--;
  return withoutQuery.slice(0, end);
}

/**
 * Determines whether a path is one of the suppressed health endpoints.
 * @param path The request path, with or without a query string.
 * @returns True when the normalized path matches a suppressed endpoint exactly.
 */
export function isSuppressedPath(path: string | undefined): boolean {
  if (path === undefined || path === "") return false;
  const normalized = normalize(path).toLowerCase();
  return SUPPRESSED_HEALTH_PATHS.includes(normalized);
}

/**
 * Determines whether telemetry should be suppressed for a path.
 * @param path The request path, with or without a query string.
 * @returns True when telemetry must be suppressed.
 * @remarks
 * Fails safe toward suppression: an unset or malformed env var leaves suppression
 * enabled, so cost control holds even when the environment is misconfigured. Only the
 * literal `false` re-enables health telemetry.
 */
export function shouldSuppressTelemetry(path: string | undefined): boolean {
  return parseSuppressionFlag(process.env[SUPPRESSION_ENV_VAR]) && isSuppressedPath(path);
}
